aspnetcore/release-notes/aspnetcore-3.0.md

```diff
@@ -0,0 +1,541 @@
+        }
+
+        return currentUsername.EndsWith("@jabbr.net", StringComparison.OrdinalIgnoreCase));
+    }
+}
+```
+
+In the preceding code, `DomainRestrictedRequirement` serves as a custom `IAuthorizationRequirement`. Because the `HubInvocationContext` resource parameter is being passed in, the internal logic can:
+
+* Inspect the context in which the Hub is being called.
+* Make decisions on allowing the user to execute individual Hub methods.
+
+Individual Hub methods can be decorated with the name of the policy the code checks at run-time. As clients attempt to call individual Hub methods, the `DomainRestrictedRequirement` handler runs and controls access to the methods. Based on the way the `DomainRestrictedRequirement` controls access:
+
+* All logged-in users can call the `SendMessage` method.
+* Only users who have logged in with a `@jabbr.net` email address can view users’ histories.
+* Only `bob42@jabbr.net` can ban users from the chat room.
+
+```csharp
+[Authorize]
+public class ChatHub : Hub
+{
+    public void SendMessage(string message)
+    {
+    }
+
+    [Authorize("DomainRestricted")]
+    public void BanUser(string username)
+    {
+    }
+
+    [Authorize("DomainRestricted")]
+    public void ViewUserHistory(string username)
+    {
+    }
+}
+```
+
+Creating the `DomainRestricted` policy might involve:
+
+* In *Startup.cs*, adding the new policy.
+* Provide the custom `DomainRestrictedRequirement` requirement as a parameter.
+* Registering `DomainRestricted` with the authorization middleware.
+
+```csharp
+services
+    .AddAuthorization(options =>
+    {
+        options.AddPolicy("DomainRestricted", policy =>
+        {
+            policy.Requirements.Add(new DomainRestrictedRequirement());
+        });
+    });
+```
+
+SignalR hubs use [Endpoint Routing](xref:fundamentals/routing). SignalR hub connection was previously done explicitly:
+
+```csharp
+app.UseSignalR(routes =>
+{
+    routes.MapHub<ChatHub>("hubs/chat");
+});
+```
+
+In the previous version, developers needed to wire up controllers, Razor pages, and hubs in a variety of different places. Explicit connection results in a series of nearly-identical routing segments:
+
+```csharp
+app.UseSignalR(routes =>
+{
+    routes.MapHub<ChatHub>("hubs/chat");
+});
+
+app.UseRouting(routes =>
+{
+    routes.MapRazorPages();
+});
+```
+
+SignalR 3.0 hubs can be routed via endpoint routing. With endpoint routing, typically all routing can be configured in `UseRouting`:
+
+```csharp
+app.UseRouting(routes =>
+{
+    routes.MapRazorPages();
+    routes.MapHub<ChatHub>("hubs/chat");
+});
+```
+
+ASP.NET Core 3.0 SignalR added:
+
+* Client-to-server streaming. With client-to-server streaming, server-side methods can take instances of either an `IAsyncEnumerable<T>` or `ChannelReader<T>`. In the following C# sample, the `UploadStream` method on the Hub will receive a stream of strings from the client:
```

Comment:
Feels like this doesn't need to be a list now that there's only one element. I'm OK either way though.

Reply:
```suggestion
Client-to-server streaming. With client-to-server streaming, server-side methods can take instances of either an `IAsyncEnumerable<T>` or `ChannelReader<T>`. In the following C# sample, the `UploadStream` method on the Hub will receive a stream of strings from the client:
```